### PR TITLE
Pytorch losses

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
This PR adds:

- [`Lovasz softmax`](https://arxiv.org/abs/1705.08790) 
- [`SSIM as a loss` ](https://en.wikipedia.org/wiki/Structural_similarity)
- fixes microbatch and decay

### CODE REVIEW

**Review by roman-kh — APPROVED** [review]

File: batchflow/models/torch/losses/lovasz.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+Heavily based on author's implementation: https://github.com/bermanmaxim/LovaszSoftmax
+"""
+import torch
+from torch import nn
+import torch.nn.functional as F
+from torch.autograd import Variable
+
+
+class BinaryLovaszLoss(nn.Module):
```
Comment (nikita-klsh) [inline batchflow/models/torch/losses/lovasz.py:13]: Why you didn't sort out new losses across the `binary.py` and `multiclass.py` files?

File: batchflow/models/torch/losses/ssim.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        for _ in self.WEIGHTS:
+            ssim_map, sigma_map = self.compute_ssim(prediction, target, return_sigma=True)
+            ssims.append(ssim_map.mean())
+            sigmas.append(sigma_map.mean())
+
+            prediction = self.pooling(prediction)
+            target = self.pooling(target)
+
+        output = 1
+        for sigma, weight in zip(sigmas[:4], self.WEIGHTS[:4]):
```
Comment (anton-br) [inline batchflow/models/torch/losses/ssim.py:98]: why do we consider only first four weights?

File: batchflow/models/torch/losses/ssim.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        if return_sigma:
+            return ssim_map, sigma_map
+        return ssim_map
+
+
+class MSSIM(SSIM):
+    """ Compute SSIM on multiple scales and return the weighted average.
+    Note that images must be grayscale.
+    """
+    WEIGHTS = [0.0448, 0.2856, 0.3001, 0.2363, 0.1333]
```
Comment (anton-br) [inline batchflow/models/torch/losses/ssim.py:76]: where did you take these numbers? I didn't find them in the reference paper.

File: batchflow/models/torch/losses/ssim.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    def __init__(self, kernel_size=11, sigma=1.5):
+        super().__init__()
+
+        self.pooling = nn.AvgPool2d(kernel_size=2, stride=2)
```
Comment (anton-br) [inline batchflow/models/torch/losses/ssim.py:81]: consider to use variable in forward function instead of class attribute

File: batchflow/models/torch/losses/lovasz.py (hunk 1 above)
Comment (SergeyTsimfer) [reply batchflow/models/torch/losses/lovasz.py:13]: Those loss functions are quite long 
Also, they deserve some private space

File: batchflow/models/torch/losses/ssim.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+class SSIM(nn.Module):
+    """ Structural similarity between two images.
+    Note that images must be grayscale.
+
+    Parameters
+    ----------
+    kernel_size : int
+        Size of filtering window.
+    sigma : number
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply batchflow/models/torch/losses/ssim.py:20]: `int` or `float`; saw that in `scipy` documentation

File: batchflow/models/torch/losses/ssim.py (hunk 2 above)
Comment (SergeyTsimfer) [reply batchflow/models/torch/losses/ssim.py:98]: I followed the MATLAB implementation of the authors

File: batchflow/models/torch/losses/ssim.py (hunk 4 above)
Comment (SergeyTsimfer) [reply batchflow/models/torch/losses/ssim.py:81]: I actually did it to avoid creating this layer in each of the forward passes

**Review by anton-br — APPROVED** [review]
adjusted for the last comment.

File: batchflow/models/torch/losses/core.py [hunk 6]
```diff
@@ -3,11 +3,64 @@
...
-class CrossEntropyLoss(nn.CrossEntropyLoss):
-    """ Custom loss which casts target dtype if needed """
-    def forward(self, input, target):
+class CrossEntropyLoss(nn.Module):
+    """ Custom loss which casts target dtype if needed.
+    Additionally, allows string specifiers for `weight` parameter.
+
+    Parameters
+    ----------
+    weight : str, callable or torch.Tensor
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply batchflow/models/torch/losses/core.py:14]: Done
